### ISSUE TITLE
Invalid Filter data in unused stream objects

### ISSUE THREAD

**Issue by pesco** [pdf-association/pdf-issues#216]
**Description**
The PDF specification does not make a clear statement whether the syntax rules of a stream filter must be obeyed if the stream in question is not used anywhere. The closest I can find is in section 7.4.1 "General":

> A PDF writer may encode data in a stream (for example, data for sampled images) to compress it or to convert it to a portable ASCII representation (or both). 
> A PDF reader shall invoke the corresponding decoding filter or filters to convert the information back to its original form.

The first sentence ("may encode...") lets one suspect that the second sentence applies to any stream object, used or not. The second sentence ("shall invoke...") could be taken to imply that a deviation from the filter syntax as described by the appropriate subsection of 7.4 would be a decoding error that makes the file invalid. It would be nice to have the intended interpretation made explicit.

**Context**
To elaborate, the spec does not (as far as my reading goes) prohibit a file from containing objects that are never used anywhere. For example, a presumably useless stream object like this would be valid:

    1 0 obj
    << /Length 5 >>
    stream
    !!!!!
    endstream
    endobj

Section 7.3.8 "Stream objects" introduces the Filter entry which is further described in Section 7.4 "Filters". The latter section describes the expected format for each filter. This includes some statements such as, for example, the following in section 7.4.3 "ASCII85Decode filter":

> Any other characters, and any character sequences that represent impossible
> combinations in the ASCII base-85 encoding, shall cause an error.

This statement, due to the part "shall cause an error" can be taken to imply that a file is invalid if it contains a stream object, used or unused, that lists `/Filter /ASCII85Decode` in its stream dictionary and contains such an "impossible combination".

However, there occur other statements about filter syntax that may not be quite as clear, because they do not explicitly mandate "shall cause an error". An example is also in section 7.4.3:

> As a special case, if all five bytes are 0, they shall be represented by the character 
> with code 122 (z) instead of by five exclamation points (! ! ! ! !).

So, if the above example object were still not used anywhere, but called for the ASCII85Decode filter, would the file as a whole be valid or invalid?

    1 0 obj
    << /Length 5 /Filter /ASCII85Decode >>
    stream
    !!!!!
    endstream
    endobj

**Proposed solution**
Add a sentence to the Filter entry in Table 5 (section 7.3.8 "Stream objects"):
> The stream data shall follow the appropriate format(s) as specified in 7.4, "Filters".

**Comment by lrosenthol** [comment]
> The PDF specification does not make a clear statement whether the syntax rules of a stream filter must be obeyed if the stream in question is not used anywhere.

Correct.   Because ISO 32000 isn't a document describing validation only processing (and creation).   Since such an object wouldn't be referenced, thus not read, it doesn't matter what you do with it.

> The second sentence ("shall invoke...") could be taken to imply that a deviation from the filter syntax as described by the appropriate subsection of 7.4 would be a decoding error that makes the file invalid.

Except that it is a processor requirement and not a file format requirement.  And we say anything about what a processor should do when it encounters a stream that it can't decode.  Like most things, it is specifically left as processor independent.

> So, if the above example object were still not used anywhere

If the object isn't used, then it doesn't matter.

**Comment by pesco** [comment]
> Because ISO 32000 isn't a document describing validation only processing (and creation).

I am obviously not a member of the relevant body, so can't speak authoritatively on what the authors of the spec intend, but I will in all politeness disagree with your statement, and I think your "Securing PDF" working group will support this view.

_Of course_ the specification describes validation. Implicitly, when it makes various normative ("shall") statements about the proper form of a PDF file, and explicitly when it requires PDF processors to produce an error in some case.
 
Quoting section 0.2 "ISO 32000 and PDF":
> The primary purpose of this document is to define well-formed PDF documents (conforming PDF files).

To the issue at hand: While the spec makes it clear that it intends PDF files to be read in what is sometimes called the "island" parsing strategy (random access via the cross-reference data) rather than a linear fashion, it does describe the expected form of the full file as a sequence of objects in 7.5 "File structure".

So the question remains: Is a file valid ("conforming") that contains a stream with data that is malformed with respect to its Filter?

**Comment by lrosenthol** [comment]
>I am obviously not a member of the relevant body, 

You are welcome to join us anytime...

>so can't speak authoritatively on what the authors of the spec intend, 

As someone who has been responsible for the content of that specification for almost 20 years now, I feel pretty authoritative about that my remark.

>but I will in all politeness disagree with your statement, and I think your "Securing PDF" working group will support this view.

That group was founded to consider *changing* the current position of the specification...so yes, there are those who believe that ISO 32000 should be more strict in both file format specification and processing logic.  Whether any/all of the proposals in that group will find their way into a future version is TBD...

>While the spec makes it clear that it intends PDF files to be read in what is sometimes called the "island" parsing strategy (random access via the cross-reference data) rather than a linear fashion,

Never heard that term before, but yes - a PDF *must* be processed through the cross-reference - or specifically, object data is retrieved by indexing through the cross-reference.  (there is nothing that says you can't read the actual data in any order you want - it's just about finding the objects themselves).

>it does describe the expected form of the full file as a sequence of objects in 7.5 "File structure".

True, but that again, don't confuse file format specifications with processing rules.

>Is a file valid ("conforming") that contains a stream with data that is malformed with respect to its Filter?

I would argue that such a state is "undefined", except in the specific cases where the standard calls out an explicit error - such as the malformed ASCII85 data.

**Comment by pesco** [comment]
> I would argue that such a state is "undefined"

> The primary purpose of this document is to define well-formed PDF documents

I am tired of having Internet arguments with you over what are obviously issues with the PDF specification that I am reporting on the official PDF Association PDF issues repository.

**Comment by MatthiasValvekens** [comment]
In the interest of avoiding going into the weeds on what it means to "define" something, and to avoid beating the "what is validation" dead horse: here's my 2 cents on the original question.

 - For better or worse, there's no prohibition on "garbage" data in between PDF objects, so if the would-be object has no xref pointing to it, I 'd argue that nothing is invalidated.
 - If there is a pointer in the xref table to the piece of data under scrutiny, I think it necessarily has to be a well-formed object, regardless of whether the object is used anywhere in the DOM. (=> full disclosure: I didn't check the actual wording in the spec yet, but that feels like something we should clarify if it's not in there already)
- I personally also subscribe to the converse position: any data that doesn't have an xref pointing to it is functionally not a PDF object, but there's definitely no language in the standard to back that up. It's also not really relevant to this discussion, I think.

Basically, I'd argue that any data that can be accessed from the xref table is part of the PDF object model and should therefore abide by the provisions of the standard, to the extent that a processor that supports all the relevant filters (*) should be able to decode it. Whether an object is actually used/required/... is (IMHO) besides the point.

The question then becomes: do all "referenceable" stream objects that use (standard) filters have to contain data that is valid according to the definition of those filters? I think so.

(*) Remaining wrinkle: custom filters, crypt filters and whatnot are in principle permissible. _If_ this discussion leads to a proposed spec change, then the added text should be clearly scoped to only apply to those filters that are actually defined in the standard. Otherwise it's a breaking change.

TL;DR: I think I support @pesco's conclusion, but maybe not for all the same reasons. Having said that, I also think the proposed change is too broad, but I won't try to wordsmith until we reach some kind of consensus.

**Comment by lrosenthol** [comment]
>I am tired of having Internet arguments with you

Then come join the committee and we can have them face to face...

I am glad that you feel strongly about these issues and welcome the discussions.   You must, however, keep in mind that "technical purity" isn't always the right choice over customer/business requirements.

**Comment by pesco** [comment]
> The question then becomes: do all "referenceable" stream objects that use (standard) filters have to contain data that is valid according to the definition of those filters?

For the record, this is the question as I meant it. Specifically, I didn't want to touch on what kind of "junk" is allowed between the objects that are listed in the cross-reference table(s).

> I think so.

I am tentatively of the same opinion (hence my proposed solution), but I'll also state that the primary issue I see at  hand (and I think I worded the original post as such) is that the spec does not clearly explain its own stance on it.

To @petervwyatt: I concur on all of the points you make. One may be writing a validator, viewer, or other processor, and their particular implementation choices are theirs alone, but what every one of them needs is a clear specification of what is officially considered a conforming file.

**Comment by petervwyatt** [comment]
In agreement - this would then be a (reasonable) file format requirement, not a processor requirement, since not all processors will necessarily read all streams. No one should expect a generic PDF processor to report violations on streams that it doesn't need to process! That would be a role for validators for the stricter subsets, but ISO 32000 itself does not discuss validation at all.

**Comment by petervwyatt** [comment]
Some lightweight proposed solutions to current 7.4.1 wording to try and make things crystal clear (note that this section has not been reworked by ISO since fast-track):

* "_Whether to do so and which decoding filter or filters to use ~~are~~ **shall be** specified in the stream dictionary."
* "_PDF ~~files~~ **processors shall** support a standard set of filters that fall into two main categories:_"
* Append to the end of the 1st para in 7.4.1: "_All stream data shall follow the appropriate format(s) as described below._" - not doing this in Table 5 since would need to be added twice for **Filter** and **FFilter**.

**Comment by petervwyatt** [comment]
PDF TWG agree
